Tuxedo Parachain Support (#1932)
w3f/Grants-Program: applications/tuxedo_parachain.md

### Overview

In a [previous grant](https://github.com/w3f/Grants-Program/blob/master/applications/tuxedo.md) ([PR](https://github.com/w3f/Grants-Program/pull/1517)) we developed Tuxedo, a framework for writing Substrate runtimes in the UTXO model.

Currently, Tuxedo works well for standalone Substrate blockchains, but does not yet work as a Polkadot Parachain.
In this grant, we propose to implement the common infrastructure necessary for a UTXO runtime to fulfill the parachain side of the Polkadot protocol.

We further propose to collect benchmark data profiling Tuxedo in various contexts.

### Project Details

Tuxedo is analogous to FRAME in that they are both frameworks for writing Substrate runtimes.
They are also analogous in that they need additional code to serve as parachain runtimes.

The code that allows a FRAME to work as a parachain is a large part of cumulus.
Specifically, it is in the [Parachain System pallet](https://paritytech.github.io/cumulus/cumulus_pallet_parachain_system/index.html) and the [`generate_validate_block!` macro](https://paritytech.github.io/cumulus/src/cumulus_pallet_parachain_system_proc_macro/lib.rs.html#94-153).

> The first two milestones are devoted to adapting and rewriting these two ipmortant pieces of code in a UTXO-native way.

Reviewer: ```suggestion
The first two milestones are devoted to adapting and rewriting these two important pieces of code in a UTXO-native way.

```

* Main Tuxedo Repo - https://github.com/Off-Narrative-Labs/Tuxedo
* Tuxedo Tutorial - https://github.com/Off-Narrative-Labs/Tuxedo-Order-Book-Dex-Tutorial/
* Nimbus Consensus (Prior Work) - https://github.com/moonbeam-foundation/nimbus
* Blockchain From Scratch (Educational) https://github.com/JoshOrndorff/blockchain-from-scratch

## Development Status :open_book:

As mentioned above, the team has done previous work on [Tuxedo](https://github.com/Off-Narrative-Labs/Tuxedo/tree/main).

As we mentioned in our previous grant our next tasks are parachain support (this grant's topic) and zero knowledge UTXOs.

Of course there are also some lingering leftovers in our issue queue mostly centered around the wallet.
We do at some point want to make the wallet more generalizable to be able to craft any generic transaction based on whichever tuxedo runtime is defined.

## Development Roadmap :nut_and_bolt:

### Overview

- **Total Estimated Duration:** 12 weeks
- **Full-Time Equivalent (FTE):**  1 FTE 
- **Total Costs:** $48,000 (USD)

### Milestone 1 — Rewrite Pallet Parachain System to UTXO model 

- **Estimated duration:** 6 weeks
- **FTE:**  1
- **Costs:** 24,000 USD

> The bulk of the logic for handling communication with the relay chain lives in pallet parachain system, and consequently, this is the largest milestone in this grant proposal.

Reviewer: ```suggestion
The bulk of the logic for handling communication with the relay chain lives in the pallet parachain system, and consequently, this is the largest milestone in this grant proposal.

```

I didn't really get what you mean here; "parachain system" is the name of the pallet? Or do you mean "pallet parachain system" as a single term? If so, what is the definition of that system?

Author: The name of the pallet is "Pallet Parachain System" or maybe just "Parachain System". I think this is the common way to refer to them, right? Like "Pallet Balances" or "Pallet Staking".

In any case, this is the thing I'm referring to, and this link is already included previously in the application. https://paritytech.github.io/cumulus/cumulus_pallet_parachain_system/index.html

Author: ```suggestion
The bulk of the logic for handling communication with the relay chain lives in [pallet parachain system](https://paritytech.github.io/cumulus/cumulus_pallet_parachain_system/index.html), and consequently, this is the largest milestone in this grant proposal.

```

Reviewer: Great, thanks for the clarification.

- **Costs:** 8,000 USD

The `register_validate_block!` macro is responsible for generating the `validate_block` entrypoint function in the runtime wasm. This is the main entrypoint in the runtime from the relay chain validators' perspective.

We will need to migrate the following to a Tuxedo compatible version

```rust
cumulus_pallet_parachain_system::register_validate_block! {
	Runtime = Runtime,
	BlockExecutor = cumulus_pallet_aura_ext::BlockExecutor::<Runtime, Executive>,
	CheckInherents = CheckInherents,
}
```

| Number | Deliverable | Specification |
| -----: | ----------- | ------------- |
| **0a.** | License | Apache 2.0 |
| **0b.** | Documentation | We will provide both **inline documentation** of the code and a basic **tutorial** that explains the difference between the FRAME based approach and the Tuxedo based approach to this macro |
| **0c.** | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
| **0d.** | Docker | We will provide a Dockerfile that can be used to test all the functionality delivered with this milestone. |

> | 1.  | Tuxedo Parachain Template | If you crossed the standalone Tuxedo Template with the Cumulus Template. This is the main deliverable for the grant. A working tempalte node for a Tuxedo-based parachain runtime and colaltor node. |

Reviewer: ```suggestion
| 1.  | Tuxedo Parachain Template | If you crossed the standalone Tuxedo Template with the Cumulus Template. This is the main deliverable for the grant. A working template node for a Tuxedo-based parachain runtime and collator node. |

```

> If you crossed the standalone Tuxedo Template with the Cumulus Template.

This sentence seems to be incomplete.

- **Estimated Duration:** 4 weeks
- **FTE:**  1
- **Costs:** 16,000 USD

Benchmark common transaction types in standalone and parachain contexts.
Make good faith comparisons to analogous transaction types in FRAME.
We hope that the simplicity of the UTXO model will allow performance increases, but right now that is just a theory.

| Number | Deliverable | Specification |
| -----: | ----------- | ------------- |
| **0a.** | License | Apache 2.0 |
| **0b.** | Documentation | We will provide both **inline documentation** of the code and a basic **tutorial** that explains how a user can spin up the example node and transfer tokens |
| **0c.** | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
| **0d.** | Docker | We will provide a Dockerfile that can be used to test all the functionality delivered with this milestone. |
| 1.  | Benchmark Report | Full report of throughput for {Transfer, Remark} transactions in {FRAME, Tuxedo} runtimes operated in {Standalone, Parachain} contexts. |

## Future Plans

### Near Term

> We have a vision for a Tuxedo parachain that acts as an Atomic Swap hub for DOT ecosystem assets to foreign UTXO chains like Monero, Zcash, Cardano, etc.

Reviewer: How about Bitcoin?

Reviewer: Makes sense, thanks for the clarification. 👍 
It might be interesting to compare this approach to Interlay's, though I agree that this shouldn't be the main focus here.

The [Farcaster](https://github.com/farcaster-project) provides a Monero Bitcoin atomic swap protocol.
We intend to extend this protocol to support Tuxedo chains.

> And thanks to Polkadot's XCM, this would allow atomic swaps between other DOT ecosystem asstes and foreign UTXO chains as well.

Reviewer: ```suggestion
And thanks to Polkadot's XCM, this would allow atomic swaps between other DOT ecosystem assets and foreign UTXO chains as well.

```

Following completion of this grant, the path to working Atomic Swaps would be roughly

1. Extend Farcaster to support swaps with Tuxedo.
2. XCM integration with Tuxedo for Cross-chain UTXOS.
3. Testing, auditing, etc.

### Medium Term

> Otherdreams we have for Tuxedo include:

Reviewer: ```suggestion
Other dreams we have for Tuxedo include:

```
